src/coreclr/vm/stringliteralmap.cpp

```diff
@@ -201,6 +203,15 @@ STRINGREF *StringLiteralMap::GetStringLiteral(EEStringData *pStringData, BOOL bA
         // Retrieve the string objectref from the string literal entry.
         pStrObj = pEntry->GetStringObject();
         _ASSERTE(!bAddIfNotFound || pStrObj);
+
+        if (pStrObj != nullptr && preferFrozenObjectHeap && ppPinnedString != nullptr)
+        {
+            Object* underlyingStrObj = *reinterpret_cast<Object**>(pStrObj);
+            if (GCHeapUtilities::GetGCHeap()->IsInFrozenSegment(underlyingStrObj))
```

Comment:
I was hoping that you will be able to have a bit somewhere that keeps track of whether the string is frozen or not, so that we do not have to do lookups in global tables.

If you think that it is best to check the global tables like this, the check can be done in the JIT/EE interface like you had it before - hiding it behind several layers does not buy us much.

Reply:
Do you mean a bit in the object header? (I did it in my most recent commit to this PR)

Reply:
I meant more like StringLiteralEntry (e.g. steal some bits from the refcount).